### ISSUE TITLE
Resolve module with symlinks

### ISSUE THREAD

**Issue by Cellule** [nodejs/node-v0.x-archive#18203]
I have been working for a while with link to develop multiple module or simply reuse my local version.
However, every time I link a module using peerDependencies the link just becomes very very painful.

For instance, if I have a structure like
```
pkg1@1.0.0 D:\temp\pkg1
├── lodash@3.7.0
└── pkg2@1.0.0 -> D:\temp\pkg2
```
And for whatever reason `pkg2` tries to require `lodash` then I get the error `Error: Cannot find module 'lodash'`.
However, without the link, it works perfectly.
An even greater problem is if I happened to have lodash installed globally it works with the link, but not using the version intended because I could have version 2.8.0 installed globally. 
This is worst, because you don't get an error, everything runs and at one point it might break and you will have to search everywhere to finally realise you were not using the right version because of your link.

Now I wanted to try to find the source of the problem and maybe fix it.
I realized that when you require `pkg2` in my example, its path will be `D:\temp\pkg2\index.js` instead of `D:\temp\pkg1\node_modules\pkg2\index.js` which result in the resolve paths to be
```
paths: Array[3]
0: "D:\temp\pkg2\node_modules"
1: "D:\temp\node_modules"
2: "D:\node_modules"
```
Instead of
```
paths: Array[4]
0: "D:\temp\pkg1\node_modules\pkg2\node_modules"
1: "D:\temp\pkg1\node_modules"
2: "D:\temp\node_modules"
3: "D:\node_modules"
```

I tried to change the function tryFile in module.js as follow
```js
// check if the file exists and is not a directory
function tryFile(requestPath) {
  var fs = NativeModule.require('fs');
  var stats = statPath(requestPath);
  if (stats && !stats.isDirectory()) {
    return requestPath; // <= Fix: use the symlink directly
    //return fs.realpathSync(requestPath, Module._realpathCache);
  }
  return false;
}
```
And it works.

I do believe there is a good reason why it resolves the symlink at this point, but I don't know it.
Maybe a better solution would be to use the parent's modules path as well when resolving modules.

**Comment by SystemParadox** [comment]
Absolutely need to fix this and use the correct virtual path for symlinks instead of resolving them. The current behaviour makes `npm link` horrific to work with.

This is the cause of https://github.com/npm/npm/issues/5080

**Comment by Cellule** [comment]
For the time being I solved this issue with

``` js
var res = require.resolve("module-x");
if(res.indexOf("node_modules") === -1) {
  try { 
    require("app-module-path").addPath(require("path").join(__dirname, "node_modules"));
  } catch(e) {  }
}
```

The module [`app-module-path`](https://www.npmjs.com/package/app-module-path) "hacks" Node's Module resolution which is obviously not the best solution for this use case.
Since this is only a problem in development, it is not the end of the world, but having a solution from Node would improve the experience when working with symlinks.

**Comment by Cellule** [comment]
Any update on this? Again today I encountered a problem with this using `npm link`.
I was using (codependency)[https://www.npmjs.com/package/codependency] to verify if an optional peer dependency is installed. 
However, it keeps failing because it uses my module real path instead of the link.
It searches for `moduleDep` around `C:\moduleA` instead of `C:\moduleB\node_modules\moduleA`
because I made an `npm link` in `moduleB` to `moduleA` so
`C:\moduleB\node_modules\moduleA` `=>` `C:\moduleA`.
